liberum_core/src/swarm_runner/messages.rs

```diff
@@ -134,18 +160,65 @@ impl SwarmContext {
                     .insert(query_id, response_sender);
                 Ok(false)
             }
+
+            // Download a file from a given peer
             SwarmRunnerMessage::DownloadFile {
                 id,
                 peer,
                 response_sender,
             } => {
+                println!(
+                    "Sending a file_share request for id {} to peer {}",
+                    bs58::encode(id.clone()).into_string(),
+                    peer.to_base58()
+                );
+
+                // If the local peer
+                if &peer == self.swarm.local_peer_id() {
+                    // Should be implemented using a VAULT
+                    let file = self.behaviour.providing.get(&id);
+                    if let Some(file) = file {
+                        match file {
+                            file_share::SharedResource::File { path } => {
+                                if let Ok(data) = tokio::fs::read(path).await {
+                                    let _ = response_sender.send(Ok(data));
+                                    return Ok(false);
+                                }
+                            }
+                        }
+                    }
+                } else {
+                    // Send a request to the peer
+                    let qid = self
+                        .swarm
+                        .behaviour_mut()
+                        .file_share
+                        .send_request(&peer, file_share::FileRequest { id: id.to_vec() });
+
+                    self.behaviour
+                        .pending_download_file
+                        .insert(qid, response_sender);
+                }
+                Ok(false)
+            }
+
+            SwarmRunnerMessage::PublishFile {
+                record,
+                response_sender,
+            } => {
+                debug!("Publishing file {:?}", record.key);
+
+                //self.swarm.behaviour_mut().kademlia.start_providing(record.key.clone())?;
+                //self.put_record_into_vault(record.clone());
+
                 let qid = self
                     .swarm
                     .behaviour_mut()
-                    .file_share
-                    .send_request(&peer, file_share::FileRequest { id: id.to_vec() });
+                    .kademlia
+                    .put_record(record, kad::Quorum::One)?; // at least one other node MUST store to consider it successful
```

Comment:
One may be too low I think...

Reply:
That's the absolute minimum of other nodes that should accept the request to provide a file to call it a success. If there are only two nodes in the network a quorum of anything more than one would fail. It's not like only one node will be asked to provide. up to K=20 nodes will be asked to do that. The quorum only decides if the query will be treated as a success or a failure. In a mature network there probably would be no reason to leave it at One though? At this moment it needs to stay low for the tests to work.

Reply:
In normal network it should probably be more, so that kad will try to send more and more requests until quorum is satisfied.